### PR TITLE
libibumad: Add support for SA MCMemberRecord

### PR DESCRIPTION
**Pull request by hnrose**
Add umad_sa_mcm.h to support SA MCMemberRecord
and add umad_sa_mcm_rereg_test.c (rewrite of infiniband-diags umad_rereg_test) using this new header.
Update umad_sa.h with some common enums, defines, and helpers shared with SA PathRecord and MultiPathRecord.

### CODE REVIEW

File: libibumad/tests/umad_sa_mcm_rereg_test.c [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
```
Comment (jgunthorpe) [inline libibumad/tests/umad_sa_mcm_rereg_test.c]: Drop HAVE_CONFIG_H it is old

File: libibumad/tests/umad_sa_mcm_rereg_test.c [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+#define MAX_PORT_GUIDS 64
+
+void build_user_mad_addr(uint8_t *umad);
+void build_mcm_rec(struct umad_sa_packet *sa, uint8_t method,
+		   const uint8_t mgid[], const uint8_t port_gid[],
+		   uint64_t tid, int create);
+
+/* Use null MGID to request SA assigned MGID */
+static uint8_t null_mgid[16] = {
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
```
Comment (jgunthorpe) [inline libibumad/tests/umad_sa_mcm_rereg_test.c]: static const

**Comment by hnrose** [comment]
@jgunthorpe 

Finally got it down to 1 error which I don't understand:

2119-----------------------------------------------------------------------------------

2120/tmp/tmp.3iQWBBx3cc/0001-libibumad-Add-umad_sa_mcm.h-for-umad-SA-MCMemberReco.patch

2121-----------------------------------------------------------------------------------

2122ERROR: Macros with complex values should be enclosed in parentheses

2123#72: FILE: libibumad/umad_sa_mcm.h:41:

2124+#  define END_C_DECLS   }

2125

2126total: 1 errors, 0 warnings, 204 lines checked

2127

2128NOTE: For some of the reported defects, checkpatch may be able to

2129      mechanically convert to the typical style using --fix or --fix-inplace.

2130

2131/tmp/tmp.3iQWBBx3cc/0001-libibumad-Add-umad_sa_mcm.h-for-umad-SA-MCMemberReco.patch has style problems, please review.

2132------------------------------------------------------------------------------

Line 41 in umad_sa_mcm.h is same as in many other files (as is line 40) for BEGIN_C_DECLS and END_C_DECLS.

Any idea what it doesn't like ? Why does it think this is "complex macro" ?

**Comment by jgunthorpe** [comment]
That one is bogus, we can ignore it.

However, IMHO we should never use the BEGIN_C_DECLS approach in public header files, they should all directly open code the '#ifdef __cplusplus' to avoid symbol pollution.

**Comment by hnrose** [comment]
@jgunthorpe 
You wrote:
IMHO we should never use the BEGIN_C_DECLS approach in public header files, they should all directly open code the '#ifdef __cplusplus' to avoid symbol pollution.

It's in libibverbs/verbs.h and others in rdma-core.

How does open coding this eliminate symbol pollution ? Wouldn't it effectively be the same thing if those macros are replaced with the open coding ? Does the difference when !cplusplus somehow matter ? If so, I'll submit another patch to open code all such occurrences in rdma-core.

**Comment by jgunthorpe** [comment]
BEGIN_C_DECLS is the symbol that creates the pollution. We shouldn't define it or use it, as the user may be using it for something else.

Changing all files would be great.

**Comment by hnrose** [comment]
I just fixed up umad_sa_mcm.h to open code the extern "C" when __cplusplus is defined. I'll do the others in a separate patch.

**Comment by jgunthorpe** [comment]
Looks fine now

**Comment by rleon** [comment]
@hnrose please rebase.

Thanks

**Comment by hnrose** [comment]
@rleon done as requested; baking now...

**Pull request merged**